pkg/node-labels-resources/grpc_server.go

```diff
@@ -0,0 +1,339 @@
+
+	"github.com/liqotech/liqo/pkg/consts"
+	resourcemonitors "github.com/liqotech/liqo/pkg/liqo-controller-manager/resource-request-controller/resource-monitors"
+	"github.com/liqotech/liqo/pkg/utils"
+)
+
+type NodeDetails struct {
+	Schedulable bool
+	Allocatable corev1.ResourceList
+	Pods        map[string]corev1.ResourceList
+}
+
+type NodeLabelsMonitor struct {
+	Server *grpc.Server
+	resourcemonitors.ResourceReaderServer
+	subscribers   sync.Map
+	nodeLabels    map[string]string
+	k8sNodeClient v1.NodeInterface
+	k8sPodClient  v1.PodInterface
+	allocatable   corev1.ResourceList
+	nodeMutex     sync.RWMutex
+	ctx           context.Context
+	resourceLists map[string]NodeDetails
+}
+
+func ListenAndServeGRPCServer(port int, nodeLabels map[string]string, clientset *kubernetes.Clientset) error {
+	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	resyncPeriod := 10 * time.Hour
+	ctx := ctrl.SetupSignalHandler()
+	if err != nil {
+		return fmt.Errorf("failed to listen on port %d: %w", port, err)
+	}
+
+	// this function is used to filter and ignore virtual nodes at informer level.
+	var noVirtualNodesFilter = func(options *metav1.ListOptions) {
+		labelSelector := labels.NewSelector()
+		for k, v := range nodeLabels {
+			req, err := labels.NewRequirement(k, selection.Equals, []string{v})
+			utilruntime.Must(err)
+			labelSelector = labelSelector.Add(*req)
+		}
+		req, err := labels.NewRequirement(consts.TypeLabel, selection.NotEquals, []string{consts.TypeNode})
+		utilruntime.Must(err)
+		labelSelector = labelSelector.Add(*req)
+		klog.V(1).Infof("Node selector resource monitor label selector: %s", labelSelector.String())
+		options.LabelSelector = labelSelector.String()
+	}
+
+	// this function is used to filter and ignore shadow pods at informer level.
+	var noShadowPodsFilter = func(options *metav1.ListOptions) {
+		req, err := labels.NewRequirement(consts.LocalPodLabelKey, selection.NotEquals, []string{consts.LocalPodLabelValue})
+		utilruntime.Must(err)
+		options.LabelSelector = labels.NewSelector().Add(*req).String()
+		options.FieldSelector = fields.OneTermEqualSelector("status.phase", string(corev1.PodRunning)).String()
+	}
+
+	nodeFactory := informers.NewSharedInformerFactoryWithOptions(
+		clientset, resyncPeriod, informers.WithTweakListOptions(noVirtualNodesFilter),
+	)
+	nodeInformer := nodeFactory.Core().V1().Nodes().Informer()
+	podFactory := informers.NewSharedInformerFactoryWithOptions(
+		clientset, resyncPeriod, informers.WithTweakListOptions(noShadowPodsFilter),
+	)
+	podInformer := podFactory.Core().V1().Pods().Informer()
+	s := NodeLabelsMonitor{
+		Server:        grpc.NewServer(),
+		nodeLabels:    nodeLabels,
+		allocatable:   corev1.ResourceList{},
+		k8sNodeClient: clientset.CoreV1().Nodes(),
+		k8sPodClient:  clientset.CoreV1().Pods(corev1.NamespaceAll),
+		ctx:           ctx,
+		resourceLists: map[string]NodeDetails{},
+	}
+	nodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Comment:
```suggestion
	utilruntime.Must(nodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
```

The same for the other occurrence